### ISSUE TITLE
Our hash function for floats that are ints doesn't match python

### ISSUE THREAD

**Issue by braxtonmckee** [APrioriInvestments/typed_python#142]
Python has a behavior where `hash(1.0)` == `hash(1)` == `hash(True)` == 1. This happens so that you can look up an int in a dictionary with a float and vice versa (since they also compare equal). We are definitely not matching this behavior, and so (a) we will allow a Dict(OneOf(float, int)) to have separate values for 1.0 and 1, which python does not allow, and (b) we will consider {1:2} to be != {1.0: 2}.

**Comment by szymonlipinski** [comment]
It looks like there is something else checked too https://stackoverflow.com/questions/114830/is-a-python-dictionary-an-example-of-a-hash-table/33459086#33459086

**Comment by braxtonmckee** [comment]
yes, python looks in has bucket first, then equality of items using __eq__.

**Comment by braxtonmckee** [comment]
We should also not be comparing things that are different types like this unless we can actually dispatch to an operator properly.  Python won't let you check ("hi",) < (1.0,), (although it'll let you check ==). We should be mimicking that behavior.
